### PR TITLE
Implement experimental filesystem

### PR DESCRIPTION
**Pull request by sir-hassan**
## Summery

This Draft PR adds a simple filesystem in database core package.

### CODE REVIEW

File: pkg/refactor/filestore/psql/root.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	if res.RowsAffected != 1 {
+		return fmt.Errorf("unexpedted gorm delete count, got: %d, want: %d", res.RowsAffected, 1)
+	}
+
+	return nil
+}
+
+//nolint:ireturn
+func (r *Root) CreateFile(ctx context.Context, path string, typ filestore.FileType, dataReader io.Reader) (filestore.File, error) {
+	path = filepath.Clean(path)
```
Comment (alankm) [inline pkg/refactor/filestore/psql/root.go:49]: Let's put an extra check/step in here to either ensure absolute paths or convert relative paths to absolute ones. That will reduce the number of possible bugs related to filepaths.

File: pkg/refactor/filestore/psql/root.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	depth := strings.Count(path, "/")
+	if path == "/" {
+		depth = 0
+	}
+
+	var data []byte
+	if dataReader != nil {
+		data, _ = io.ReadAll(dataReader)
```
Comment (alankm) [inline pkg/refactor/filestore/psql/root.go:57]: This error should not be suppressed.

File: pkg/refactor/mirror/mirror.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		if d.IsDir() {
+			_, err = direktivRoot.CreateFile(ctx, strings.TrimPrefix(path, distDir), filestore.FileTypeDirectory, nil)
+		} else {
+			_, err = direktivRoot.CreateFile(ctx, strings.TrimPrefix(path, distDir), filestore.FileTypeFile, fileReader)
+		}
+
+		if err != nil {
+			return fmt.Errorf("mirror create filestore entry, err: %w", err)
+		}
+		lg.Debugw("mirror file saved in filestore")
```
Comment (alankm) [inline pkg/refactor/mirror/mirror.go:87]: This log is both necessary but also potentially misleading. Presumably this logic should all run in a transaction. Unless we so some comprehensive validation before opening the transaction, there's a good chance the process will be aborted. Rollback is desirable, but if that happens, these logs are lies.

**Comment by alankm** [comment]
Don't forget to sign your commits. :)

File: pkg/refactor/filestore/psql/file.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	"io"
+	"path/filepath"
+	"time"
+
+	"github.com/direktiv/direktiv/pkg/refactor/filestore"
+	"github.com/google/uuid"
+	"gorm.io/gorm"
+)
+
+type File struct {
```
[the remark this answers is not in the document]
Comment (alankm) [reply pkg/refactor/filestore/psql/file.go:16]: It looks to me like gorm probably does rely on this. But we should un-export these things when we swap from gorm to sql.

**Comment by alankm** [comment]
Your exhaustive linting is throwing loads of errors for me. I think it hasn't been tried in a while?

File: pkg/refactor/filestore/psql/root.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	if path == "/" {
+		depth = 0
+	}
+
+	res := r.db.WithContext(ctx).
+		// Don't include file 'data' in the query. File data can be retrieved with file.GetData().
+		Select("id", "path", "depth", "root_id", "created_at", "updated_at").
+		Where("root_id", r.ID).
+		Where("depth", depth+1).
+		Where("path LIKE ?", path+"%", path).
```
Comment (alankm) [inline pkg/refactor/filestore/psql/root.go:109]: This line has a bug with a missing slash. I think it is also bugged just to look at: there's only one question mark, but we passed in two args. This is a mistake, right?

File: pkg/refactor/mirror/mirror.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	}()
+
+	err = source.PullInPath(settings, distDir)
+	if err != nil {
+		return fmt.Errorf("mirror pull, err: %w", err)
+	}
+
+	lg.Debugw("mirror fetched", "dist_directory", distDir)
+
+	err = filepath.WalkDir(distDir, func(path string, d fs.DirEntry, err error) error {
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/mirror/mirror.go]: That's correct. Let's leave this point open.

**Comment by sir-hassan** [comment]
Todos:

- [ ] Add pagination support.
- [ ] Add sync solution.
- [ ] Add transactional interfaces.
- [ ] Implement revisions and references.

File: pkg/refactor/core/namespace.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+// Namespace represent a core domain object of direktiv.
+type Namespace interface {
+	GetID() uuid.UUID
+	GetName() string
+	CreatedAt() time.Time
+	UpdatedAt() time.Time
+}
+
+// NamespaceStore responsible for fetching namespace from datastore.
+type NamespaceStore interface {
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/core/namespace.go:19]: Fixed.

File: pkg/refactor/filestore/filestore.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+type File interface {
+	GetID() uuid.UUID
+	GetPath() string
+	GetName() string
+	GetType() FileType
+
+	GetData(ctx context.Context) (io.ReadCloser, error)
+
+	Delete(ctx context.Context, forceDelete bool) error
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/filestore/filestore.go]: I changed the api du to drastically to enable transactions. This method has now different position and shape.

[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/filestore/filestore.go]: Fixed it :)

File: pkg/refactor/filestore/filestore.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	GetAllRoots(ctx context.Context) ([]Root, error)
+}
+
+type Root interface {
+	GetID() uuid.UUID
+	GetFile(ctx context.Context, path string) (File, error)
+
+	CreateFile(ctx context.Context, path string, typ FileType, dataReader io.Reader) (File, error)
+
+	ListPath(ctx context.Context, path string) ([]File, error)
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/filestore/filestore.go]: Renamed :)

File: pkg/refactor/filestore/psql/file.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/filestore/psql/file.go:12]: Thats correct, it is needed by **gorm** and **json marshalling**.

File: pkg/refactor/mirror/mirror.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+// and any other details).
+type Settings struct{}
+
+// Source is an interface that represent a mirror source (git repo is a valid mirror source example).
+// In Direktiv, a Mirror is a directory of files that sits somewhere (local or remote) and a user wants to mirror (copy)
+// his direktiv namespace files from it.
+// Source knows how to access the mirror files (connecting to a remote server in case of git) and copy the files in
+// the user's direktiv namespace. Parameter 'settings' is used to configure the sourcing (pulling) mirror process.
+// Parameter 'dir' specifies the directory where Source should copy the mirror to.
+type Source interface {
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/mirror/mirror.go:27]: Sure, let's leave this open.

File: pkg/refactor/mirror/mirror.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	// PullInPath pulls (copies) mirror into local directory specified by 'dir' parameter.
+	PullInPath(mirrorSettings Settings, dir string) error
+}
+
+// ExecuteMirroringProcess pulls mirror from source, store it in local file system and then push it to direktiv
+// filestore.
+func ExecuteMirroringProcess(
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/mirror/mirror.go:34]: Good case to add unit test for it. 👍🏼

File: pkg/refactor/mirror/mirror.go (hunk 3 above)
Comment (sir-hassan) [reply pkg/refactor/mirror/mirror.go:93]: Thats correct. 

Now I changed the apis of this filestore to enable transactions, individual  **info** and **debug** lines are potentially lies if the parent transaction is rolled back, but the full logging information would **not** be a lie if the caller of  caller of **filestore** did good job logging around (before and after) a transactional **filestore** operation properly - indicating the opration was commited or rolled back resulting to something like:

```
log: starting transaction to filestore.
log: debug: filestore created dir.
log: debug: filestore created file.
log: operation failed due network error.
log: rolling back filestore operation.
```

Plus, logging inside any module is typically only debug level (error should be returned to the caller controller or higher level modules) that is activated only on development system.

Does this make sense?

File: pkg/refactor/filestore/filestore.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+type Filestore interface {
+	CreateRoot(ctx context.Context, id uuid.UUID) (Root, error)
+
+	GetRoot(ctx context.Context, id uuid.UUID) (Root, error)
+	GetAllRoots(ctx context.Context) ([]Root, error)
+}
+
+type Root interface {
+	GetID() uuid.UUID
+	GetFile(ctx context.Context, path string) (File, error)
```
[the remark this answers is not in the document]
Comment (alankm) [reply pkg/refactor/filestore/filestore.go]: We've decided not to do this. We should create a separate function to have separate behaviour.

File: pkg/refactor/filestore/filestore.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
+package filestore
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/filestore/filestore.go:1]: Fixed.

File: pkg/refactor/filestore/psql/root.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+
+	return f, nil
+}
+
+// nolint:ireturn
+func (r *Root) ListPath(ctx context.Context, path string) ([]filestore.File, error) {
+	var list []File
+
+	path = filepath.Clean(path)
+	depth := strings.Count(path, "/")
```
[the remark this answers is not in the document]
Comment (sir-hassan) [reply pkg/refactor/filestore/psql/root.go]: Fixed.

**Pull request closed without merging**